src/tools/pkcs11-tool.c

```diff
@@ -4050,25 +4048,43 @@ static CK_RV write_object(CK_SESSION_HANDLE session)
 	memset(&gost,  0, sizeof(gost));
 #endif
 
-	memset(contents, 0, sizeof(contents));
-	memset(certdata, 0, sizeof(certdata));
-
 	f = fopen(opt_file_to_write, "rb");
 	if (f == NULL)
 		util_fatal("Couldn't open file \"%s\"", opt_file_to_write);
-	contents_len = fread(contents, 1, sizeof(contents) - 1, f);
+	if (fseek(f, 0L, SEEK_END) != 0)
+		util_fatal("Couldn't set file position to the end of the file \"%s\"", opt_file_to_write);
+	contents_len = ftell(f);
 	if (contents_len < 0)
+		util_fatal("Couldn't get file position \"%s\"", opt_file_to_write);
+	contents = malloc(contents_len + 1);
+	if (contents == NULL)
+		util_fatal("malloc() failure\n");
+	if (fseek(f, 0L, SEEK_SET) != 0)
+		util_fatal("Couldn't set file position to the beginning of the file \"%s\"", opt_file_to_write);
+	size_t ret = fread(contents, 1, contents_len, f);
```

Comment:
I would prefer to have the declaration at the beginning of the block. Sounds like no CI we run enforces this, but it is a good practice to avoid shadowing the variable declarations (such as you do on line 4084) and getting unexpected errors.